enum.rb

```diff
@@ -0,0 +1,122 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    i = 0
+    range = is_a?(Range) ? to_a : self
+    par = is_a?(Range) ? size : length
+    while i < par
+      if is_a?(Hash)
+        yield(keys[i], self[keys[i]])
+      else
+        yield(range[i])
+      end
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    my_each do
+      if is_a?(Hash)
+        yield(self[keys[i]], keys[i])
+      else
+        yield(self[i], i)
+      end
+      i += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr = []
+    my_each { |i| arr << i if yield(i) }
+    arr
+  end
+
+  def my_all?
+    my_each do |i|
+      if !arg[0].nil?
+        return false unless arg[0].equal? i
+      elsif block_given?
+        return false unless yield(i)
+      else
+        return false unless i
+      end
+    end
+    true
+  end
+
+  def my_any?(*arg)
+    my_each do |i|
+      if !arg[0].nil?
+        return true if arg[0].equal? i
+      elsif block_given?
+        return true if yield(i)
+      elsif i
+        return true
+      end
+    end
+    false
+  end
+
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  def my_count(arg = nil)
+    count = 0
+    my_each do |i|
+      if !arg.nil?
+        count += 1 if i == arg
+      elsif block_given?
+        count += 1 if yield(i)
+      else
+        count += 1
+      end
+    end
+    count
+  end
+
+  def my_map(proc = nil)
+    return to_enum unless block_given? or proc
+
+    arr = []
+    my_each do |i|
+      arr << block_given? ? yield(i) : proc.call(i)
+    end
+    arr
+  end
```

Comment:
#### my_map

- [ ] accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

Given 
`my_proc = Proc.new {|num| num > 10 }`

- Yours is giving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/76906403-7d398d80-6869-11ea-8d77-e97eed83ee47.png)

- Real
![image](https://user-images.githubusercontent.com/52209067/76906442-8cb8d680-6869-11ea-9f6d-d864f9cef00b.png)

[OPTIONAL]

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend making it work

Given 
`my_proc = Proc.new {|num| num > 10 }`

Result
> puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  # true true false false